In [1]:
import pandas as pd, numpy as np, matplotlib.pyplot as plt, seaborn as sns

In [2]:
from sklearn.model_selection import train_test_split

In [3]:
# Load all 4 datasets 
company_info = pd.read_csv("company_info.csv")
market_indices = pd.read_csv("market_indices.csv")
stock_prices = pd.read_csv("stock_prices.csv")
stock_prices_indicators = pd.read_csv("stock_prices_with_indicators.csv")
print (" All CSVc loaded successfully (if no errors shown).")

 All CSVc loaded successfully (if no errors shown).


In [4]:
datasets = {
    "company_info": company_info,
    "market_indices": market_indices,
    "stock_prices": stock_prices,
    "stock_prices_indicators": stock_prices_indicators
}

### 1. Company_info dataset 

In [5]:
company_info.shape

(20, 4)

In [6]:
company_info.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20 entries, 0 to 19
Data columns (total 4 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   ticker        20 non-null     object
 1   company_name  20 non-null     object
 2   sector        20 non-null     object
 3   ipo_date      20 non-null     object
dtypes: object(4)
memory usage: 772.0+ bytes


In [7]:
company_info.head()

ticker company_name      sector    ipo_date
0  STK001     TechCorp  Technology  2021-04-10
1  STK002  DataSystems  Technology  2016-10-12
2  STK003    CloudNine  Technology  2016-01-18
3  STK004  CyberShield  Technology  2022-02-21
4  STK005    MediPharm  Healthcare  2018-03-05

#### Insight 
Company_Info dataset has 20 rows and 4 columns with no null values. This file help to map tickers to company names and sectors for grouping, filtering, and sector-based analysis.

### 2. Market_Indices dataset

In [8]:
market_indices.shape

(780, 7)

In [9]:
market_indices.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 780 entries, 0 to 779
Data columns (total 7 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   date           780 non-null    object 
 1   sp500_close    780 non-null    float64
 2   nasdaq_close   780 non-null    float64
 3   vix_close      780 non-null    float64
 4   treasury_10y   780 non-null    float64
 5   dollar_index   780 non-null    float64
 6   market_regime  780 non-null    object 
dtypes: float64(5), object(2)
memory usage: 42.8+ KB


In [10]:
market_indices.head()

date  sp500_close  nasdaq_close  vix_close  treasury_10y  \
0  2021-01-04      3891.04      12952.02      20.48         1.496   
1  2021-01-05      3847.20      13116.52      19.89         1.504   
2  2021-01-06      3878.24      12818.48      20.02         1.507   
3  2021-01-07      3803.95      13290.11      19.59         1.498   
4  2021-01-08      3715.64      13365.73      20.14         1.498   

   dollar_index market_regime  
0         90.07          bull  
1         90.73          bull  
2         90.46          bull  
3         90.67          bull  
4         90.64          bull

#### Insight 
Market Indices has 780 rows and 7 columns with no null entires 

### 3. Stock_Prices dataset 

In [11]:
stock_prices.shape

(15600, 8)

In [12]:
stock_prices.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15600 entries, 0 to 15599
Data columns (total 8 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   ticker          15600 non-null  object 
 1   date            15600 non-null  object 
 2   open            15600 non-null  float64
 3   high            15600 non-null  float64
 4   low             15600 non-null  float64
 5   close           15600 non-null  float64
 6   volume          15600 non-null  int64  
 7   adjusted_close  15600 non-null  float64
dtypes: float64(5), int64(1), object(2)
memory usage: 975.1+ KB


In [13]:
stock_prices.head()

ticker        date    open    high     low   close   volume  adjusted_close
0  STK001  2021-01-04  158.09  160.97  158.09  160.11   962644          160.11
1  STK001  2021-01-05  163.16  165.50  160.76  162.36  1312685          162.36
2  STK001  2021-01-06  161.89  162.51  160.94  161.78  1449177          161.78
3  STK001  2021-01-07  163.33  167.90  163.33  167.07  1534833          167.07
4  STK001  2021-01-08  168.20  168.20  164.12  165.68   848261          165.68

#### Insight 
Stock_prices dataset has 15600 rows and 8 columns with no null entires 

### 4. Stock_Prices_Indicators

In [14]:
stock_prices_indicators.shape

(15502, 31)

In [15]:
stock_prices_indicators.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15502 entries, 0 to 15501
Data columns (total 31 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   ticker            15502 non-null  object 
 1   date              15502 non-null  object 
 2   open              15502 non-null  float64
 3   high              15502 non-null  float64
 4   low               15502 non-null  float64
 5   close             15502 non-null  float64
 6   volume            15502 non-null  float64
 7   adjusted_close    15502 non-null  float64
 8   sma_20            15192 non-null  float64
 9   sma_50            15502 non-null  float64
 10  sma_200           15502 non-null  float64
 11  ema_12            15502 non-null  float64
 12  ema_26            15502 non-null  float64
 13  macd              15191 non-null  float64
 14  macd_signal       15502 non-null  float64
 15  macd_histogram    15502 non-null  float64
 16  rsi_14            15172 non-null  float6

In [16]:
stock_prices_indicators.head()

ticker        date    open    high     low   close     volume  \
0  STK001  2021-01-04  158.09  160.97  158.09  160.11   962644.0   
1  STK001  2021-01-05  163.16  165.50  160.76  162.36  1312685.0   
2  STK001  2021-01-06  161.89  162.51  160.94  161.78  1449177.0   
3  STK001  2021-01-07  163.33  167.90  163.33  167.07  1534833.0   
4  STK001  2021-01-08  168.20  168.20  164.12  165.68   848261.0   

   adjusted_close      sma_20      sma_50  ...  true_range  atr_14  \
0          160.11  160.110000  160.110000  ...        2.88   2.880   
1          162.36  161.235000  161.235000  ...        5.39   4.135   
2          161.78  161.416667  161.416667  ...        1.57   3.280   
3          167.07  162.830000  162.830000  ...        6.12   3.990   
4          165.68  163.400000  163.400000  ...        4.08   4.008   

   volume_sma_20  volume_ratio  momentum_10  momentum_20  price_to_sma_50  \
0      962644.00      1.000000          NaN          NaN         0.000000   
1     1137664.50      1.153842          NaN          NaN         0.006977   
2     1241502.00      1.167277          NaN          NaN         0.002251   
3     1314834.75      1.167320          NaN          NaN         0.026039   
4     1221520.00      0.694431          NaN          NaN         0.013953   

   volatility_20  future_return_5d  trend_label  
0            NaN          0.026357      Uptrend  
1            NaN          0.018169     Sideways  
2       0.012463         -0.012301     Sideways  
3       0.018138         -0.061950    Downtrend  
4       0.018663         -0.070196    Downtrend  

[5 rows x 31 columns]

#### Insight 
stock_prices_indicators has 15502 rows, 31 columns and some null values

In [18]:
total_missing = stock_prices_indicators.isnull().sum().sum()
total_cells = stock_prices_indicators.shape[0] * stock_prices_indicators.shape[1]
overall_missing_percent = (total_missing / total_cells) * 100

print(f"Overall missing percentage: {overall_missing_percent:.2f}%")


Overall missing percentage: 0.47%


In [19]:
# Identify datatype mismatches 
for name, df in datasets.items():
    print(f"\n📘 Data Types in {name}")
    display(df.dtypes)



📘 Data Types in company_info


ticker          object
company_name    object
sector          object
ipo_date        object
dtype: object


📘 Data Types in market_indices


date              object
sp500_close      float64
nasdaq_close     float64
vix_close        float64
treasury_10y     float64
dollar_index     float64
market_regime     object
dtype: object


📘 Data Types in stock_prices


ticker             object
date               object
open              float64
high              float64
low               float64
close             float64
volume              int64
adjusted_close    float64
dtype: object


📘 Data Types in stock_prices_indicators


ticker               object
date                 object
open                float64
high                float64
low                 float64
close               float64
volume              float64
adjusted_close      float64
sma_20              float64
sma_50              float64
sma_200             float64
ema_12              float64
ema_26              float64
macd                float64
macd_signal         float64
macd_histogram      float64
rsi_14              float64
bb_middle           float64
bb_upper            float64
bb_lower            float64
bb_width            float64
true_range          float64
atr_14              float64
volume_sma_20       float64
volume_ratio        float64
momentum_10         float64
momentum_20         float64
price_to_sma_50     float64
volatility_20       float64
future_return_5d    float64
trend_label          object
dtype: object

In [20]:
# Fixing object type for date columns 
for name, df in datasets.items():
    # Convert 'date' column if it exists
    if 'date' in df.columns:
        try:
            df['date'] = pd.to_datetime(df['date'], errors='coerce')
            print(f" Converted 'date' to datetime in {name}")
        except Exception as e:
            print(f" Error converting 'date' in {name}: {e}")
    
    # Convert 'ipo_date' column if it exists
    if 'ipo_date' in df.columns:
        try:
            df['ipo_date'] = pd.to_datetime(df['ipo_date'], errors='coerce')
            print(f" Converted 'ipo_date' to datetime in {name}")
        except Exception as e:
            print(f" Error converting 'ipo_date' in {name}: {e}")


 Converted 'ipo_date' to datetime in company_info
 Converted 'date' to datetime in market_indices
 Converted 'date' to datetime in stock_prices
 Converted 'date' to datetime in stock_prices_indicators


In [21]:
# Check for duplicate 
for name, df in datasets.items():
    duplicates = df.duplicated().sum()
    print(f"{name}: {duplicates} duplicate rows")


company_info: 0 duplicate rows
market_indices: 0 duplicate rows
stock_prices: 0 duplicate rows
stock_prices_indicators: 0 duplicate rows


In [22]:
# The percentage of missing value per row is less than 2% so we will drop them
stock_prices_indicators_clean = stock_prices_indicators.dropna()
print(f"✅ Dropped missing rows. Remaining: {len(stock_prices_indicators_clean)}")


✅ Dropped missing rows. Remaining: 13653


### Insight: 
The datasets contained no duplicate records. Missing values were detected only in the stock_prices_indicators dataset, which is expected due to the rolling nature of the technical indicators. Since the missing values accounted for less than 2% of the data, the affected rows were dropped. Additionally, the date column was converted to Pandas datetime format for consistency.